### PR TITLE
RD-5955 Update UI packages NPM publish mechanism

### PR DESCRIPTION
**Pull request by qooban**
## Description

Updated `create-version.sh` script and `Jenkinsfile` to allow sub-packages publishing.

Applied updated script in other repositories:
* https://github.com/cloudify-cosmo/cloudify-ui-components/pull/167
* https://github.com/cloudify-cosmo/cloudify-blueprint-topology/pull/227

As right now `cloudify-ui-common` repository will host 5 NPM packages, so release notes mechanism must be also updated. Example of a new way for providing release notes:
* [cloudify-ui-common releases / scripts_v0.0.2-pre.1](https://github.com/cloudify-cosmo/cloudify-ui-common/releases/tag/scripts_v0.0.2-pre.1)

## Screenshots / Videos
N/A

## Checklist
- [x] My code follows the [UI Code Style](https://cloudifysource.atlassian.net/l/c/x8fq902p).
- [x] I followed the [UI Testing Guidelines](https://cloudifysource.atlassian.net/l/cp/udCVfvrx) and verified that all tests/checks have passed.
- [x] I verified if that change requires any update in the [documentation](https://cloudifysource.atlassian.net/l/c/4XKtPocy).
- [x] I added proper labels to this PR.

## Tests
Successfully published few versions of `cloudify-ui-common-scripts` package - see https://www.npmjs.com/package/cloudify-ui-common-scripts.

## Documentation
Updated main `README.md` file.

### CODE REVIEW

File: jenkins/Jenkinsfile [hunk 1]
```diff
@@ -58,7 +58,7 @@ pipeline {
     }
     stage('Deploy') {
       when {
-        expression { return env.BRANCH =~ /^publish-v.*/ }
+        expression { return env.BRANCH =~ /^publish_.*/ }
```
Comment (qooban) [inline jenkins/Jenkinsfile:61]: As right now versions can be like `4.5.2-pre.4` I used `_` as a separator here.

File: package.json [hunk 2]
```diff
@@ -11,9 +11,7 @@
     "author": "Cloudify Platform Ltd. <cosmo-admin@cloudify.co>",
     "scripts": {
         "clean": "git clean -fXd -e \\!node_modules -e \\!node_modules/**/*",
-        "publish:major": "bash ci/publish-version.sh major",
-        "publish:minor": "bash ci/publish-version.sh minor",
-        "publish:patch": "bash ci/publish-version.sh patch"
+        "publish": "bash packages/scripts/create-version.sh"
```
Comment (qooban) [inline package.json:14]: I think it's OK to simplify it and rely on user.
It's easy to use as if you call `npm run publish` you'll see help description for `create-version.sh` script.

File: packages/backend/package.json [hunk 3]
```diff
@@ -2,7 +2,7 @@
     "name": "cloudify-ui-common-backend",
     "version": "0.0.1",
     "description": "Common Cloudify UI backend library",
-    "homepage": "https://github.com/cloudify-cosmo/cloudify-ui-common/blob/master/backend/README.md",
+    "homepage": "https://github.com/cloudify-cosmo/cloudify-ui-common/blob/master/packages/backend/README.md",
```
Comment (qooban) [inline packages/backend/package.json:5]: Not related, but as I spotted that after publishing prerelease version of `cloudify-ui-common-scripts` in NPM (https://www.npmjs.com/package/cloudify-ui-common-scripts) I decided to fix all packages at once.

File: packages/configs/package.json [hunk 4]
```diff
@@ -17,6 +17,7 @@
         "eslint": "^7.32.0",
         "eslint-config-airbnb": "^18.2.1",
         "eslint-config-airbnb-base": "^14.2.1",
+        "eslint-config-prettier": "^6.12.0",
```
Comment (qooban) [inline packages/configs/package.json:20]: Not related, but it was indeed missing here and blocking me from proceeding, so decided to fix it at once.

File: packages/scripts/create-version.sh [hunk 5]
```diff
@@ -1,86 +1,137 @@
...
-      false
-    }
-    npm run build || {
-      log "Running build failed"
-      false
-    }
+VERSION_TYPE=$1
+PACKAGE_NAME=$2
 
-    NEXT_VERSION=$(npm version --no-git-tag-version "$@")
```
Comment (qooban) [inline packages/scripts/create-version.sh:59]: `npm version` works differently for workspaces (not producing single line output with version), so I couldn't use it in all cases.

File: packages/scripts/create-version.sh [hunk 6]
```diff
@@ -1,86 +1,137 @@
...
+  false
+}
+npm run build --if-present || {
+  error "Running build failed"
+  false
+}
+
+NPM_VERSION_OPTIONS="--no-git-tag-version $VERSION_TYPE"
+if [[ $VERSION_TYPE == pre* ]]; then
+    NPM_VERSION_OPTIONS="$NPM_VERSION_OPTIONS --preid pre"
```
Comment (qooban) [inline packages/scripts/create-version.sh:93]: Decided to go with this kind of prefix. It's used when the following version types are used: `prepatch`, `preminor`, `premajor`, `prerelease`, and produces something like: `1.2.3-pre.0`

File: packages/scripts/create-version.sh [hunk 7]
```diff
@@ -1,86 +1,137 @@
...
+else
+    NEXT_VERSION_BRANCH="publish_${NEXT_VERSION}"
+    COMMIT_MESSAGE="Bump version to ${NEXT_VERSION}"
+    GIT_TAG="${NEXT_VERSION}"
+fi
+
+git checkout -b ${NEXT_VERSION_BRANCH}
+git add package.json
+if [ -n "$IS_WORKSPACE_PACKAGE" ]; then
+    git add ../../package-lock.json
```
Comment (qooban) [inline packages/scripts/create-version.sh:114]: That's needed as root package `package-lock.json` file changes after changing version of sub-package.

File: packages/scripts/create-version.sh [hunk 8]
```diff
@@ -1,86 +1,137 @@
...
+  false
+}
+npm run test --if-present || {
+  error "Running tests failed"
+  false
+}
+npm run build --if-present || {
+  error "Running build failed"
+  false
+}
```
Comment (kubama) [inline packages/scripts/create-version.sh:89]: We have `set -e` at the beginning of the script. So what's the purpose of these conditional constructs?

**Comment by Vorbert-Kruk** [comment]
As I'm having a day off tomorrow, I'm opting out as a reviewer to not block this PR from progressing
If my input would still be desirable on Monday, I'll be more than glad to provide it :)

Sorry for any inconvenience

File: README.md [hunk 9]
```diff
@@ -78,13 +78,14 @@ This section is divided into two parts:
...
-
-    * `npm run publish:patch` for new patch version,
-    * `npm run publish:minor` for new minor version,
-    * `npm run publish:major` for new major version,
-
-   which will create special branch, add commit to it containing version bump in `package*.json` files according to your choice, tag the commit and push branch to remote. That should trigger Jenkins jobs finalizing publish.
+1. Provide version type and package name to `npm run publish`. For example:
+
+    * `npm run publish -- minor cypress` - to publish new minor version of `cloudify-ui-common-cypress` package
+    * `npm run publish -- prepatch scripts` - to publish new prerelease patch version of `cloudify-ui-common-scripts`
```
[the remark this answers is not in the document]
Comment (qooban) [reply README.md]: Right, good point. Removed.

File: packages/scripts/create-version.sh (hunk 8 above)
Comment (qooban) [reply packages/scripts/create-version.sh:89]: I think it's here to add user-friendly error log.

**Review by kubama — APPROVED** [review]

**Review by PatrycjaKubiczek — DISMISSED** [review]
LGTM

File: packages/scripts/create-version.sh (hunk 8 above)
Comment (qooban) [reply packages/scripts/create-version.sh]: Personally I think it helps in some cases, so I would leave it.

Example:
```
Your branch is up to date with 'origin/RD-5955'.
Already up to date.
 packages/backend/test/logger.test.ts | 2 +-
 packages/scripts/create-version.sh   | 6 ++----
 2 files changed, 3 insertions(+), 5 deletions(-)
create-version.sh: ERROR: Working directory not clean. Delete or push your changes to RD-5955.
```

IMHO without "Working directory not clean. Delete or push your changes to RD-5955." it wouldn't be clear what to do next.

I'm just adding small fix - https://github.com/cloudify-cosmo/cloudify-ui-common/pull/139/commits/ff9823cc22a210989bed8eab41160ccd3b1a7ff9 .

**Review by kubama — APPROVED** [review]

**Pull request merged**